File: 5. Longest Palindromic Substring.cpp
//Solution1
class Solution {
public:
    string longestPalindrome(string s) {
        const int size_s = s.size();
        int max_s = 0, max_l = 0;
        for (int i = 0; i < size_s;) {
            int start = i, end = i;
            while (end + 1 < size_s && s[end] == s[end + 1]) end++;
            i = end + 1;
            while (start - 1 >= 0 && end + 1 < size_s && s[start - 1] == s[end + 1]) {
                start--;
                end++;
            }
            if (end - start + 1 > max_l) {
                max_l = end - start + 1;
                max_s = start;
            }
        }
        return s.substr(max_s, max_l);
    }
};

//Solution2
string longestPalindrome(string s) {
    if (s.empty()) return "";
    string prep = "#";
    for (auto ch : s) {
        prep += ch;
        prep += "#";
    }
    const int size_p = prep.size();
    vector<int> dp(size_p, 0);
    int center = 0, bCur = 0;
    for (int i = 0; i < size_p; i++) {
        int mirror = center - (i - center);
        dp[i] = bCur <= i ? 0 : min(bCur - i, dp[mirror]);
        int start = i - dp[i], end = i + dp[i];
        while (start - 1 >= 0 && end + 1 < size_p && prep[start - 1] == prep[end + 1]) {
            --start;
            ++end;
            ++dp[i];
        }
        if (i + dp[i] > bCur) {
            bCur = i + dp[i];
            center = i;
        }
    }
    center = max_element(dp.begin(), dp.end()) - dp.begin();
    return s.substr((center - dp[center])/2, dp[center]);
}

File: java/5. Longest Palindromic Substring.java
// O(n^2)
class Solution {
    public String longestPalindrome(String s) {
        String res = "";
        int len = 0;
        for (int i = 0; i < s.length(); ++i) {
            int j = i, k = i;
            while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
                --j;
                ++k;
            }
            if (k - j - 1 > len) {
                len = k - j - 1;
                res = s.substring(j + 1, k);
            }
            j = i;
            k = i + 1;
            while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
                --j;
                ++k;
            }
            if (k - j - 1 > len) {
                len = k - j - 1;
                res = s.substring(j + 1, k);
            }
        }
        return res;
    }
}


// dp O(n^2)
class Solution {
    public String longestPalindrome(String s) {
        int n = s.length();
        String res = null;
        boolean[][] dp = new boolean[n][n];
        for (int i = n - 1; i >= 0; --i) {
            for (int j = i; j < n; ++j) {
                dp[i][j] = s.charAt(i) == s.charAt(j) && (j - i < 3 || dp[i + 1][j - 1]);
                if (dp[i][j] && (res == null || j - i + 1 > res.length())) {
                    res = s.substring(i, j + 1);
                }
            }
        }
        return res;
    }
}
